fhevm/tee_comparison_test.go

```diff
@@ -0,0 +1,378 @@
+	op := func(lhs, rhs uint64) bool {
+		return lhs == rhs
+	}
+	signature := "teeEq(uint256,uint256,bytes1)"
+
+	testcases := []struct {
+		typ tfhe.FheUintType
+		lhs uint64
+		rhs uint64
+	}{
+		{tfhe.FheUint4, 2, 1},
+		{tfhe.FheUint8, 2, 1},
+		{tfhe.FheUint16, 4283, 1337},
+		{tfhe.FheUint32, 1333337, 1337},
+		{tfhe.FheUint64, 13333377777777777, 133377777777},
+	}
+	for _, tc := range testcases {
+		t.Run(fmt.Sprintf("teeEq with %s", tc.typ), func(t *testing.T) {
+			teeComparison1Helper(t, tc.typ, tc.lhs, tc.rhs, op, signature)
+		})
+	}
+}
+
+func TestTeeGeRun(t *testing.T) {
+	op := func(lhs, rhs uint64) bool {
+		return lhs >= rhs
+	}
+	signature := "teeGe(uint256,uint256,bytes1)"
+
+	testcases := []struct {
+		typ tfhe.FheUintType
+		lhs uint64
+		rhs uint64
+	}{
+		{tfhe.FheUint4, 2, 1},
+		{tfhe.FheUint8, 2, 1},
+		{tfhe.FheUint16, 4283, 1337},
+		{tfhe.FheUint32, 1333337, 1337},
+		{tfhe.FheUint64, 13333377777777777, 133377777777},
+	}
+	for _, tc := range testcases {
+		t.Run(fmt.Sprintf("teeGe with %s", tc.typ), func(t *testing.T) {
+			teeComparison1Helper(t, tc.typ, tc.lhs, tc.rhs, op, signature)
+		})
+	}
+}
+
+func TestTeeGtRun(t *testing.T) {
+	op := func(lhs, rhs uint64) bool {
+		return lhs > rhs
+	}
+	signature := "teeGt(uint256,uint256,bytes1)"
+
+	testcases := []struct {
+		typ tfhe.FheUintType
+		lhs uint64
+		rhs uint64
+	}{
+		{tfhe.FheUint4, 2, 1},
+		{tfhe.FheUint8, 2, 1},
+		{tfhe.FheUint16, 4283, 1337},
+		{tfhe.FheUint32, 1333337, 1337},
+		{tfhe.FheUint64, 13333377777777777, 133377777777},
+	}
+	for _, tc := range testcases {
+		t.Run(fmt.Sprintf("teeGt with %s", tc.typ), func(t *testing.T) {
+			teeComparison1Helper(t, tc.typ, tc.lhs, tc.rhs, op, signature)
+		})
+	}
+}
+
+func TestTeeNeRun(t *testing.T) {
+	op := func(lhs, rhs uint64) bool {
+		return lhs != rhs
+	}
+	signature := "teeNe(uint256,uint256,bytes1)"
+
+	testcases := []struct {
+		typ tfhe.FheUintType
+		lhs uint64
+		rhs uint64
+	}{
+		{tfhe.FheUint4, 2, 1},
+		{tfhe.FheUint8, 2, 1},
+		{tfhe.FheUint16, 4283, 1337},
+		{tfhe.FheUint32, 1333337, 1337},
+		{tfhe.FheUint64, 13333377777777777, 133377777777},
+	}
+	for _, tc := range testcases {
+		t.Run(fmt.Sprintf("teeNe with %s", tc.typ), func(t *testing.T) {
+			teeComparison1Helper(t, tc.typ, tc.lhs, tc.rhs, op, signature)
+		})
+	}
+}
+
+func TestTeeMinRun(t *testing.T) {
+	op := func(lhs, rhs uint64) uint64 {
+		if lhs <= rhs {
+			return lhs
+		} else {
+			return rhs
+		}
+	}
+	signature := "teeMin(uint256,uint256,bytes1)"
+
+	testcases := []struct {
+		typ tfhe.FheUintType
+		lhs uint64
+		rhs uint64
+	}{
+		{tfhe.FheUint4, 2, 1},
+		{tfhe.FheUint8, 2, 1},
+		{tfhe.FheUint16, 4283, 1337},
+		{tfhe.FheUint32, 1333337, 1337},
+		{tfhe.FheUint64, 13333377777777777, 133377777777},
+	}
+	for _, tc := range testcases {
+		t.Run(fmt.Sprintf("teeMin with %s", tc.typ), func(t *testing.T) {
+			teeComparison2Helper(t, tc.typ, tc.lhs, tc.rhs, op, signature)
+		})
+	}
+}
```

Comment:
@lovenoble I'm looking at these tests, and I know I wrote the first ones (copying Zama's), but I don't feel super comfortable that they are useful. I propose to put the `expected` values in the struct too. 

Do you see the difference? Do you agree it's a better test? If yes, let's do this for all the tests.

```suggestion
func TestTeeMinRun(t *testing.T) {
	signature := "teeMin(uint256,uint256,bytes1)"

	testcases := []struct {
		typ tfhe.FheUintType
		lhs uint64
		rhs uint64
		expected uint64
	}{
		{tfhe.FheUint4, 2, 1, 1},
		{tfhe.FheUint8, 2, 1, 1},
		{tfhe.FheUint16, 4283, 1337, 1337},
		{tfhe.FheUint32, 1333337, 1337, 1337},
		{tfhe.FheUint64, 13333377777777777, 133377777777, 133377777777},
	}
	for _, tc := range testcases {
		t.Run(fmt.Sprintf("teeMin with %s", tc.typ), func(t *testing.T) {
			teeComparison2Helper(t, tc.typ, tc.lhs, tc.rhs, tc.expected op, signature)
		})
	}
}
```

Reply:
okay

Reply:
If we use zama tests, there is not need to calculate the results separately in development stage.

Reply:
I agree it's more convenient. But I also think that's the reason it's less useful.

For example, for `teeSub` on uint8, we want to test that `32 - 33 == 255`. Is that possible the current format?

Reply:
I think that's a good point. 👍